File: app/src/main/java/me/imzack/app/end/view/dialog/TypeMarkColorPickerDialogFragment.java
package me.imzack.app.end.view.dialog;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.TextView;
import android.widget.ViewAnimator;

import me.imzack.app.end.App;
import me.imzack.app.end.R;
import me.imzack.app.end.model.bean.TypeMarkColor;
import me.imzack.app.end.util.ColorUtil;
import me.imzack.app.end.util.ResourceUtil;
import me.imzack.app.end.view.adapter.TypeMarkColorGridAdapter;
import me.imzack.app.end.view.widget.CircleColorView;
import me.imzack.app.end.view.widget.ColorPicker;

import java.io.Serializable;
import java.util.List;

import butterknife.BindView;

public class TypeMarkColorPickerDialogFragment extends BaseDialogFragment {

    @BindView(R.id.switcher_color_picker)
    ViewAnimator mColorPickerSwitcher;
    @BindView(R.id.grid_type_mark_color)
    GridView mTypeMarkColorGrid;
    @BindView(R.id.picker_type_mark_color)
    ColorPicker mTypeMarkColorPicker;
    @BindView(R.id.text_type_mark_color)
    TextView mTypeMarkColorText;
    @BindView(R.id.ic_type_mark_color)
    CircleColorView mTypeMarkColorIcon;

    private static final String ARG_DEFAULT_COLOR = "default_color";
    private static final String ARG_TYPE_MARK_COLOR_PICKED_LSNR = "type_mark_color_picked_lsnr";

    private String mDefaultColor;
    private TypeMarkColor mTypeMarkColor;
    private int mPosition = -1;
    private List<TypeMarkColor> mTypeMarkColorList;
    private OnTypeMarkColorPickedListener mOnTypeMarkColorPickedListener;

    public TypeMarkColorPickerDialogFragment() {

    }

    public static TypeMarkColorPickerDialogFragment newInstance(String defaultColor, OnTypeMarkColorPickedListener listener) {
        TypeMarkColorPickerDialogFragment fragment = new TypeMarkColorPickerDialogFragment();
        Bundle args = new Bundle();
        args.putString(ARG_TITLE_STR, ResourceUtil.getString(R.string.title_dialog_fragment_type_mark_color_picker));
        args.putString(ARG_NEU_BTN_STR, ResourceUtil.getString(R.string.btn_custom));
        args.putString(ARG_NEG_BTN_STR, ResourceUtil.getString(R.string.button_cancel));
        args.putString(ARG_POS_BTN_STR, ResourceUtil.getString(R.string.button_select));
        args.putString(ARG_DEFAULT_COLOR, defaultColor);
        args.putSerializable(ARG_TYPE_MARK_COLOR_PICKED_LSNR, listener);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        if (args != null) {
            mDefaultColor = args.getString(ARG_DEFAULT_COLOR);
            mOnTypeMarkColorPickedListener = (OnTypeMarkColorPickedListener) args.getSerializable(ARG_TYPE_MARK_COLOR_PICKED_LSNR);
        }

        mTypeMarkColorList = App.getDataManager().getTypeMarkColorList();

        mPosition = getPositionInTypeMarkColorList(mDefaultColor);

        mTypeMarkColor = new TypeMarkColor(
                mDefaultColor,
                mPosition == -1 ? mDefaultColor : mTypeMarkColorList.get(mPosition).getColorName()
        );

        setNeutralButtonClickListener(new OnButtonClickListener() {
            @Override
            public boolean onClick() {
                mColorPickerSwitcher.showNext();
                if (mColorPickerSwitcher.getCurrentView().getId() == R.id.grid_type_mark_color) {
                    //切换到了grid界面，此时mPosition一定为-1
                    mPosition = getPositionInTypeMarkColorList(mTypeMarkColor.getColorHex());
                    if (mPosition != -1) {
                        //若picker界面选中的颜色在grid界面也有，选中它
                        mTypeMarkColorGrid.setItemChecked(mPosition, true);
                    }
                    setNeutralButtonString(getString(R.string.btn_custom));
                } else {
                    //切换到了picker界面
                    if (mPosition != -1) {
                        //若之前在grid界面有选择，取消选择
                        mTypeMarkColorGrid.setItemChecked(mPosition, false);
                        mPosition = -1;
                    }
                    //此时mPosition一定为-1
                    mTypeMarkColorText.setText(mTypeMarkColor.getColorHex());
                    mTypeMarkColorIcon.setFillColor(Color.parseColor(mTypeMarkColor.getColorHex()));
                    mTypeMarkColorPicker.setColor(Color.parseColor(mTypeMarkColor.getColorHex()));
                    setNeutralButtonString(getString(R.string.btn_preset));
                }
                return false;
            }
        });
        setPositiveButtonClickListener(new OnButtonClickListener() {
            @Override
            public boolean onClick() {
                mPosition = getPositionInTypeMarkColorList(mTypeMarkColor.getColorHex());
                //若position为-1，说明颜色在grid中不存在，是在picker中设置的，无名称，用hex代替；反之说明颜色在grid中存在，有名称
                mTypeMarkColor.setColorName(mPosition == -1 ? mTypeMarkColor.getColorHex() : mTypeMarkColorList.get(mPosition).getColorName());
                if (mOnTypeMarkColorPickedListener != null) {
                    mOnTypeMarkColorPickedListener.onTypeMarkColorPicked(mTypeMarkColor);
                }
                return true;
            }
        });
    }

    @Override
    public View onCreateContentView(LayoutInflater inflater, ViewGroup root) {
        return inflater.inflate(R.layout.dialog_fragment_type_mark_color_picker, root, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mTypeMarkColorGrid.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                mTypeMarkColorGrid.getViewTreeObserver().removeOnPreDrawListener(this);
                //运行时设置宽度（在xml文件中宽度设置成match_parent不行）
                mTypeMarkColorPicker.getLayoutParams().width = mTypeMarkColorGrid.getWidth();
                return false;
            }
        });

        mTypeMarkColorGrid.setAdapter(new TypeMarkColorGridAdapter(mTypeMarkColorList));

        if (mPosition != -1) {
            mTypeMarkColorGrid.setItemChecked(mPosition, true);
        }

        mTypeMarkColorGrid.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                mPosition = position;
                mTypeMarkColor.setColorHex(mTypeMarkColorList.get(mPosition).getColorHex());
            }
        });

        mTypeMarkColorPicker.setOnColorChangedListener(new ColorPicker.OnColorChangedListener() {
            @Override
            public void onColorChanged(int color) {
                String colorHex = ColorUtil.parseColor(color);
                mTypeMarkColorIcon.setFillColor(color);
                mTypeMarkColorText.setText(colorHex);
                mTypeMarkColor.setColorHex(colorHex);
            }
        });
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mOnTypeMarkColorPickedListener = null;
    }

    private int getPositionInTypeMarkColorList(String color) {
        for (int i = 0; i < mTypeMarkColorList.size(); i++) {
            if (mTypeMarkColorList.get(i).getColorHex().equals(color)) {
                return i;
            }
        }
        //未在TypeMarkColorList中找到
        return -1;
    }

    public interface OnTypeMarkColorPickedListener extends Serializable {
        void onTypeMarkColorPicked(TypeMarkColor typeMarkColor);
    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/dialog/TypeMarkColorPickerDialogFragment.kt
package me.imzack.app.end.view.dialog

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.AdapterView
import kotlinx.android.synthetic.main.dialog_fragment_type_mark_color_picker.*
import me.imzack.app.end.R
import me.imzack.app.end.model.DataManager
import me.imzack.app.end.model.bean.TypeMarkColor
import me.imzack.app.end.util.ColorUtil
import me.imzack.app.end.util.ResourceUtil
import me.imzack.app.end.view.adapter.TypeMarkColorGridAdapter
import me.imzack.lib.basedialogfragment.BaseDialogFragment

class TypeMarkColorPickerDialogFragment : BaseDialogFragment() {

    companion object {

        private val ARG_DEFAULT_COLOR = "default_color"

        fun newInstance(defaultColor: Int, typeMarkColorPickedListener: (typeMarkColor: TypeMarkColor) -> Boolean): TypeMarkColorPickerDialogFragment {
            val fragment = TypeMarkColorPickerDialogFragment()
            val args = Bundle()
            putBaseArguments(
                    args,
                    ResourceUtil.getString(R.string.title_dialog_fragment_type_mark_color_picker),
                    ResourceUtil.getString(R.string.btn_custom),
                    object : OnButtonClickListener {
                        override fun onClick(): Boolean {
                            fragment.onClickSwitcher()
                            return false
                        }
                    },
                    ResourceUtil.getString(android.R.string.cancel),
                    object : OnButtonClickListener {
                        override fun onClick() = true
                    },
                    ResourceUtil.getString(R.string.button_select),
                    object : OnButtonClickListener {
                        override fun onClick() = typeMarkColorPickedListener(fragment.typeMarkColor)
                    },
                    false
            )
            args.putInt(ARG_DEFAULT_COLOR, defaultColor)
            fragment.arguments = args
            return fragment
        }
    }

    // 不提供编程来改变颜色，因为无此需求，即只有通过用户改变颜色
    // TODO 颜色0值是否有效？
    private val defaultColor by lazy { arguments!!.getInt(ARG_DEFAULT_COLOR) }
    private val typeMarkColorList = DataManager.typeMarkColorList
    private var position = -1
    private val typeMarkColor by lazy { if (position == -1) TypeMarkColor(ColorUtil.parseColor(defaultColor)) else typeMarkColorList[position].copy() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // defaultColor 被初始化，position 被更新为正确的初始值
        updatePosition(defaultColor)
    }

    override fun onCreateContentView(inflater: LayoutInflater, root: ViewGroup) =
            inflater.inflate(R.layout.dialog_fragment_type_mark_color_picker, root, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        grid_type_mark_color.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                grid_type_mark_color.viewTreeObserver.removeOnPreDrawListener(this)
                // 运行时设置宽度（在 xml 文件中宽度设置成 match_parent 不行）
                picker_type_mark_color.layoutParams.width = grid_type_mark_color.width
                return false
            }
        })

        grid_type_mark_color.adapter = TypeMarkColorGridAdapter(typeMarkColorList)

        if (position != -1) {
            grid_type_mark_color.setItemChecked(position, true)
        }

        // 预置颜色
        grid_type_mark_color.onItemClickListener = AdapterView.OnItemClickListener { _, _, pos, _ ->
            // 更新 position
            position = pos
            // 更新 typeMarkColor
            val (colorHex, colorName) = typeMarkColorList[position]
            typeMarkColor.hex = colorHex
            typeMarkColor.name = colorName
        }

        // 自定义颜色
        picker_type_mark_color.mOnColorChangedListener = {
            // it 为整型颜色值（colorInt），colorHex 为十六进制颜色值
            val colorHex = ColorUtil.parseColor(it)
            // 更新颜色预览图标和颜色文本
            ic_type_mark_color.setFillColor(it)
            text_type_mark_color.text = colorHex
            // 更新 position，如果在自定义颜色中刚好调到某个预置颜色，则 position 为预置颜色中的位置，否则为 -1
            updatePosition(it)
            // 更新 typeMarkColor
            typeMarkColor.hex = colorHex
            // 若 position 为 -1，说明颜色是在自定义颜色中设置的，无名称，用 colorHex 代替；反之说明颜色在预置颜色中存在，有名称
            typeMarkColor.name = if (position == -1) colorHex else typeMarkColorList[position].name
        }
    }

    private fun onClickSwitcher() {
        switcher_color_picker.showNext()
        val colorHex = typeMarkColor.hex
        val color = Color.parseColor(colorHex)
        if (switcher_color_picker.currentView.id == R.id.grid_type_mark_color) {
            //切换到了grid界面，此时position一定为-1
            updatePosition(color)
            if (position != -1) {
                //若picker界面选中的颜色在grid界面也有，选中它
                grid_type_mark_color.setItemChecked(position, true)
            }
            neutralButtonText = getString(R.string.btn_custom)
        } else {
            //切换到了picker界面
            if (position != -1) {
                //若之前在grid界面有选择，取消选择
                grid_type_mark_color.setItemChecked(position, false)
                position = -1
            }
            //此时position一定为-1
            text_type_mark_color.text = colorHex
            ic_type_mark_color.setFillColor(color)
            picker_type_mark_color.setColor(color)
            neutralButtonText = getString(R.string.btn_preset)
        }
    }

    /** 更新 position，如果 color 在预置颜色中，position 设置为预置颜色中的位置，否则设置为 -1 */
    private fun updatePosition(color: Int) {
        position = typeMarkColorList.indices.firstOrNull { typeMarkColorList[it].hex == ColorUtil.parseColor(color) } ?: -1
    }
}
